proposals/v3/server.md

```diff
@@ -0,0 +1,230 @@
+CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
+0008048166ff        synse-server        0.04%               48.28MiB / 1.952GiB   2.42%               788B / 0B           0B / 0B             7
+```
+
+Which is approximately a 2.64MiB difference in mem usage, the vast majority of which is
+from the in-memory caching of the aforementioned device info. At the time of this writing,
+the emulator provides 12 devices, so approximately 220kB of cached data is being kept per
+device. This will add up as potentially hundreds of devices are added to the system.
+
+#### Transaction Caching
+v3 enables [batch writes](writes.md#group-writes) with the [tag](tags.md) based routing
+system. As part of this, the notion of *write transaction* changes such that there is a
+single *primary transaction* for a write via the Synse API. Each primary transaction
+is composed of one or more *component transactions*, which represent a single device write.
+
+The component transaction is the internal transaction state generated by the plugin and
+cached in Synse Server. This effectively does not change from v2. 
+
+The primary transaction is an additional layer of abstraction on top of component transaction(s)
+which allows:
+- component transaction grouping (e.g. batch writes)
+- user-defined transaction IDs (e.g. if application is managing transaction on its own)
+
+Just as in v2, the plugin will track the component transactions and will make that data
+available to Synse Server for caching. If Synse Server terminates, a new instance can
+rebuild the cache from the plugins. If a plugin fails, the data is lost.
+
+In v3, the primary transaction info will only live in Synse Server. All transactions have
+a TTL in the system, so they are never persisted. If Synse Server terminates, the
+primary transaction information will be lost. 
+
+Synse Server will **not** provide any means for transaction persistence, at least for the
+initial release of Synse v3. It is up to the caller to handle cases where a query for
+a transaction ID returns in a "not found" error.
+
+See the API doc for more details on [write](api.md#write) and [transaction](api.md#transaction)
+behavior.
+
+### Websocket Support
+As an alternative to the current method of interfacing with Synse Server (HTTP), support can
+be added to configure it to use [websockets](https://tools.ietf.org/html/rfc6455). This can be
+useful for different deployment scenarios.
+
+While these updates would need to be reflected in any [API clients](api-clients.md), the
+change here is relatively straightforward since [websocket support is build into the Sanic
+framework](https://sanic.readthedocs.io/en/latest/sanic/websocket.html). Adding this support
+will require updates to how the server is initialized, but should generally be low-impact.
+
+The transport option will default to HTTP for backwards compatibility. It can be set via
```

Comment:
> As an alternative to the current method of interfacing with Synse Server (HTTP)

> The transport option will default to HTTP for backwards compatibility

```yaml
transport: 
  protocol: websocket
```

> The valid options for `protocol` will be: `websocket`, `http`. The default value is `http`.

@edaniszewski This is the block that suggested to me that HTTP and Websockets are mutually exclusive. It appears from this section that you must choose one or the other.

Reply:
Ah, yes. I wrote this section up before I did some experiments with websockets in Sanic so I don't think I had a clear picture on how the implementation of it actually worked at the time. This is definitely misleading/confusing, I'll update. Thanks for pointing this out!